### ISSUE TITLE
flutter web async http requests never returns

### ISSUE THREAD

**Issue by jyhnnhyj** [flutter/flutter#48678]
I'm evaluating flutter for web - have a relatively sample POC project that calls REST APIs
Sometime http.post async calls never return - but when I run the same app in iOS simulator or MacOS desktop app it works as expected - which makes me suspect something wrong with flutter web here
I have seen similar questions posted about http requests in flutter and in most cases it was an error on the api side - but here I'm certain that my backend is okay and even the flutter app works when running in macos or ios 
I tried dev, beta and master branches and in all of them I have the same issue.
When running in Chrome, I can see in networks section of the browser that the http call is made, but somehow Flutter doesn't get that...
(the code is as simple as `http.post(...)`)
Any suggestions?

**Comment by jonahwilliams** [comment]
Please include the output of `flutter doctor -v` as well as a snippet of code that reproduces the issue.

**Comment by jyhnnhyj** [comment]
flutter doctor doesn't show any issues - code snippet:
```
.
.
.
onPressed: () async {
            print('pressed');
            var url = 'https://myawesomeapi.com/whatsit/create';
            var response = await http.post(url, body: {'name': 'doodle', 'color': 'blue'});
            print('Response status: ${response.statusCode}');
            print('Response body: ${response.body}');
          },
.
.
.
```
on iOS or MacOS, runs fine, but in web browser the line after `var response = ...` never executes...

**Comment by jonahwilliams** [comment]
@jyhnnhyj  you need to include the output of flutter doctor so we know what version of flutter you are using.

**Comment by jyhnnhyj** [comment]
sure, here is the output:
```
[✓] Flutter (Channel beta, v1.12.13+hotfix.6, on Mac OS X 10.15.2 19C57, locale en-GB)
    • Flutter version 1.12.13+hotfix.6 at /Users/jyhnnhyj/development/flutter
    • Framework revision 18cd7a3601 (5 weeks ago), 2019-12-11 06:35:39 -0800
    • Engine revision 2994f7e1e6
    • Dart version 2.7.0


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/jyhnnhyj/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3, Build version 11C29
    • CocoaPods version 1.8.4

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 42.1.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] VS Code (version 1.41.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.7.1

[✓] Connected device (2 available)
    • Chrome     • chrome     • web-javascript • Google Chrome 79.0.3945.117
    • Web Server • web-server • web-javascript • Flutter Tools

• No issues found!
```

and as I mentioned in the original post, I have the same issue on dev and master as well

**Comment by yjbanov** [comment]
@jyhnnhyj Can you wrap that code in a `try/catch` and see if there are any exceptions? It is possible that the `post` is throwing, perhaps due to a CORS error, and so the next line does not execute.

Also, just to be sure is `http` imported from `package:http`? If so, you might want to file a bug there (if it's an actual bug).

**Comment by yjbanov** [comment]
Sorry, by "there" I mean https://github.com/dart-lang/http

**Comment by jyhnnhyj** [comment]
thanks for the comments - it's not an exception and I already verified that 
as I mentioned, it works when I run as an iOS app or MacOS app... (and my backend is CORS aware etc)
My http package is `import 'package:http/http.dart' as http;`
what confuses me is that it only fails in web browser... do I still need to file a bug in dart-lang/http repo?

**Comment by fmollea** [comment]
Hi, I have the same problem. I use `http.get(...)` and the request never returns. The Android app working correctly, but in web not.
I use `package:http` and my` flutter doctor -v` return:

> [✓] Flutter (Channel beta, v1.13.6, on Mac OS X 10.13.6 17G6030, locale en-AR)
>     • Flutter version 1.13.6 at /Users/fmollea/flutter
>     • Framework revision 659dc8129d (hace 3 semanas), 2019-12-30 09:24:47 -0800
>     • Engine revision bdc9708d23
>     • Dart version 2.8.0 (build 2.8.0-dev.0.0 c547f5d933)
> 
> 
> [!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
>     • Android SDK at /Users/fmollea/Library/Android/sdk
>     • Android NDK at /Users/fmollea/Library/Android/sdk/ndk-bundle
>     • Platform android-29, build-tools 28.0.3
>     • ANDROID_HOME = /Users/fmollea/Library/Android/sdk
>     • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
>     • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
>     ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses
> 
> [!] Xcode - develop for iOS and macOS (Xcode 10.1)
>     • Xcode at /Applications/Xcode.app/Contents/Developer
>     • Xcode 10.1, Build version 10B61
>     ✗ Flutter requires a minimum Xcode version of 10.2.0.
>       Download the latest version or update via the Mac App Store.
>     • CocoaPods version 1.7.5
> 
> [✓] Chrome - develop for the web
>     • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome
> 
> [✓] Android Studio (version 3.4)
>     • Android Studio at /Applications/Android Studio.app/Contents
>     • Flutter plugin version 39.0.1
>     • Dart plugin version 183.6270
>     • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
> 
> [!] IntelliJ IDEA Ultimate Edition (version 2018.1.8)
>     • IntelliJ at /Applications/IntelliJ IDEA.app
>     ✗ Flutter plugin not installed; this adds Flutter specific functionality.
>     ✗ Dart plugin not installed; this adds Dart specific functionality.
>     • For information about installing plugins, see
>       https://flutter.dev/intellij-setup/#installing-the-plugins
> 
> [✓] VS Code (version 1.41.1)
>     • VS Code at /Applications/Visual Studio Code.app/Contents
>     • Flutter extension version 3.7.1
> 
> [✓] Connected device (2 available)
>     • Chrome     • chrome     • web-javascript • Google Chrome 79.0.3945.130
>     • Web Server • web-server • web-javascript • Flutter Tools
> 
> ! Doctor found issues in 3 categories.

Any idea? Someone have the same problem? Any fix?

**Comment by vsmenon** [comment]
> My http package is import 'package:http/http.dart' as http;
what confuses me is that it only fails in web browser... do I still need to file a bug in dart-lang/http repo?

Yes, please do.  That package has web-specific and native-specific code.

**Comment by jodymac** [comment]
I also have the same issue and I am running the latest on beta channel.  Mac app, ios work just fine.  The exact same code running locally and as a final product on a webserver all return nothing when making calls to my php pages.  The app is still running, but nothing executes in the code block after the http.get.  I see many threads on this in different areas will all the same results, ask for flutter doctor results, do the try catch and then the thread just seems to fade into inactivity.

**Comment by MazEbeid** [comment]
I am facing the same issue

**Comment by yjbanov** [comment]
Is there an issue over in https://github.com/dart-lang/http/issues that this can be deduplicated into?

**Comment by herrytco** [comment]
I experience the same bug. I analyzed the request in the backend-debugger as well and it looks like that no request-body is sent along. I tried this with http as well as with Dio in this form: 

Response response = await Dio().post(
      endpoint,
      queryParameters: request.toMap(),
      data: "some dummy data to see if it even reaches the other side.",
);

The backend does not receive the request body.

**Comment by chandrikaReddyBunga** [comment]
hi , in flutter web app from http post method not getting response,

String url = Constants.getPaymentAPi;
  // Map<String, String> headers = {"Content-type": "application/json"};
  var orderID=generateRandom();
  String json = '{"ORDER_ID": "$orderID","CUST_ID": "chandrika421@gmail.com","CHANNEL_ID": "WEB","TXN_AMOUNT": "1"}';
  Response response = await post(url,body:json);
  int statusCode = response.statusCode;
  print('payment statuscode $statusCode');
  String body = response.body;
  print('payment body $body');

**Comment by xencoder** [comment]
The same issue. only happened in web browser

**Comment by harithany** [comment]
Same issue

**Comment by xencoder** [comment]
@jodymac 
Thanks. Problem got solved by adding Access-Control-Allow-Origin "*" to header.

**Comment by natebosch** [comment]
This looks like browser issues unrelated to dart or `package:http`. Has anyone looked in the chrome devtools network tab to see what the request looks like?

**Comment by JaiMistry** [comment]
For me, I ran into a similar issue and ended up using a [CORS proxy](https://github.com/Rob--W/cors-anywhere/) which I could append to the beginning of my API endpoint URL. You can do something like this:
```dart
String proxy = 'https://cors-anywhere.herokuapp.com/';
var url = 'https://myawesomeapi.com/whatsit/create';
var response = await http.post('$proxy$url', body: {'name': 'doodle', 'color': 'blue'});
```

**Comment by fareezMaple** [comment]
Same issue here.
![image](https://user-images.githubusercontent.com/60868965/89990591-2e2b4380-dcb5-11ea-9e50-568a5d17ecee.png)

**Comment by natebosch** [comment]
@fareezMaple - have you checked the browser dev tools to see if there is information about the request that might indicate why it never completed?

**Comment by fareezMaple** [comment]
﻿```
Access to fetch at 'https://api.azanpro.com/times/today.json?zone=sgr01' from origin 'https://malaysia-waktu-solat.web.app' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

The FetchEvent for "https://api.azanpro.com/times/today.json?zone=sgr01" resulted in a network error response: the promise was rejected.
flutter_service_worker.js:1 
Uncaught (in promise) TypeError: Failed to fetch
js_helper.dart:1129 Uncaught Invalid argument: Instance of 'minified:nq'
    at Object.c (https://malaysia-waktu-solat.web.app/main.dart.js:2142:3)
    at e.N (https://malaysia-waktu-solat.web.app/main.dart.js:19740:47)
    at https://malaysia-waktu-solat.web.app/main.dart.js:39998:10
    at SI.a (https://malaysia-waktu-solat.web.app/main.dart.js:3509:72)
    at SI.$2 (https://malaysia-waktu-solat.web.app/main.dart.js:20715:23)
    at S5.$2 (https://malaysia-waktu-solat.web.app/main.dart.js:20710:25)
    at QX.Mf (https://malaysia-waktu-solat.web.app/main.dart.js:21336:42)
    at QX.Me (https://malaysia-waktu-solat.web.app/main.dart.js:21338:32)
    at jL.Kd (https://malaysia-waktu-solat.web.app/main.dart.js:20833:23)
    at Pu.$0 (https://malaysia-waktu-solat.web.app/main.dart.js:20972:36)
html_dart2js.dart:18529 GET https://api.azanpro.com/times/today.json?zone=sgr01 net::ERR_FAILED
```

How do I set no-cors as stated in the dev console log?

**Comment by pedromassangocode** [comment]
Hi @jyhnnhyj 
Are you still facing this issue? If yes, can you please provide your `flutter doctor -v`, your `flutter run --verbose`?
Thank you

**Comment by SimpleChenyh** [comment]
I have the same issue only in the browser, My problem is caused by cross-domain, I use go gin framework to write a backend server, When not allowing the CORS, [example code do not allow cross-domain](https://play.golang.org/p/IP0G4NweXhL) and I got a CORS problem
![image](https://user-images.githubusercontent.com/18431131/95675542-a09f9d80-0bea-11eb-9fcb-171927e6c2d8.png)

When allowing the CORS,[example code do  allow cross-domain](https://play.golang.org/p/AM59u_-Bb55) and it's fine
![image](https://user-images.githubusercontent.com/18431131/95675592-fecc8080-0bea-11eb-9755-9e7618df08e0.png)

Maybe this is not a perfect solution, we can use a proxy like Nginx to solve this problem.

**Comment by pedromassangocode** [comment]
Without additional information, we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs, and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.

**Comment by VocsyDikshit** [comment]
Have faced the same problem while using HTTP request... and Gone through many Websites and Found a Solution, The Solution is to Enable CORS into Application. Where I don't have any Idea about, What is CORS, Can anyone Explain to me Breifly Please.
